src/rfq/protocols/bebop/client.rs

```diff
@@ -0,0 +1,411 @@
+use std::collections::{HashMap, HashSet};
+
+use async_trait::async_trait;
+use futures::{stream::BoxStream, StreamExt};
+use serde::{Deserialize, Serialize};
+use tokio_tungstenite::{connect_async_with_config, tungstenite::Message};
+use tracing;
+
+use crate::{
+    rfq::{
+        client::RFQClient,
+        errors::RFQError,
+        indicatively_priced::SignedQuote,
+        models::{GetAmountOutParams, TimestampHeader},
+    },
+    tycho_client::feed::synchronizer::{ComponentWithState, Snapshot, StateSyncMessage},
+    tycho_common::dto::{ProtocolComponent, ResponseProtocolState},
+    tycho_core::dto::Chain,
+};
+
+type BebopPriceMessage = HashMap<String, BebopPriceData>;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct BebopPriceData {
+    last_update_ts: f64,
+    /// Vec where each tuple is (price, size)
+    bids: Vec<(f64, f64)>,
+    /// Vec where each tuple is (price, size)
+    asks: Vec<(f64, f64)>,
+}
+
+#[derive(Clone)]
+pub struct BebopClient {
+    chain: Chain,
+    url: String,
+    // Pairs that we want prices for
+    pairs: HashSet<String>,
+    // Min tvl value.
+    tvl: f64,
+    // name header for authentication
+    ws_user: String,
+    // key header for authentication
+    ws_key: String,
+}
+
+fn pair_to_bebop_format(pair: &(String, String)) -> String {
+    format!("{}/{}", pair.0, pair.1)
+}
+
+impl BebopClient {
+    pub fn new(
+        chain: Chain,
+        pairs: Vec<(String, String)>,
+        tvl: f64,
+        ws_user: String,
+        ws_key: String,
+    ) -> Self {
+        let url = "wss://api.bebop.xyz/pmm/ethereum/v3/pricing".to_string();
+
+        let mut pair_names: HashSet<String> = HashSet::new();
+        for pair in pairs {
+            pair_names.insert(pair_to_bebop_format(&pair));
+        }
+        Self { url, pairs: pair_names, chain, tvl, ws_user, ws_key }
+    }
+
+    /// Calculates Total Value Locked (TVL) based on bid/ask levels.
+    ///
+    /// TVL is calculated using the formula from Bebop's documentation:
+    /// https://docs.bebop.xyz/bebop/bebop-api-pmm-rfq/rfq-api-endpoints/pricing#interpreting-price-levels
+    ///
+    /// Returns the average of bid and ask TVLs across all price levels.
+    ///
+    /// Note: This calculation assumes all pairs use the same quote token.
+    /// For cross-pair comparisons, token prices should be normalized to a common denomination.
+    fn calculate_tvl(&self, price_data: &BebopPriceData) -> f64 {
+        let bid_tvl: f64 = price_data
+            .bids
+            .iter()
+            .map(|(price, size)| price * size)
+            .sum();
+
+        let ask_tvl: f64 = price_data
+            .asks
+            .iter()
+            .map(|(price, size)| price * size)
+            .sum();
+
+        (bid_tvl + ask_tvl) / 2.0
```

Comment:
Only not sure if I should sum these tvls or avg them

Reply:
I think this ok.. this only a rough estimation. I think we don't need to be super accurate here